go-controller/pkg/ovn/policy.go

```diff
@@ -1148,276 +1159,298 @@ func hasAnyLabelSelector(peers []knet.NetworkPolicyPeer) bool {
+		}
+		txOkCallBack()
+		return nil
+	})
+	return np, err
 }
 
 // addNetworkPolicy creates and applies OVN ACLs to pod logical switch
 // ports from Kubernetes NetworkPolicy objects using OVN Port Groups
+// if addNetworkPolicy fails, create or delete operation can be retried
 func (oc *Controller) addNetworkPolicy(policy *knet.NetworkPolicy) error {
 	klog.Infof("Adding network policy %s in namespace %s", policy.Name,
 		policy.Namespace)
 
-	nsInfo, nsUnlock, err := oc.ensureNamespaceLocked(policy.Namespace, false, nil)
-	if err != nil {
-		return fmt.Errorf("unable to ensure namespace for network policy: %s, namespace: %s, error: %v",
-			policy.Name, policy.Namespace, err)
-	}
-	_, alreadyExists := nsInfo.networkPolicies[policy.Name]
-	if alreadyExists {
-		nsUnlock()
-		// If this scenario happens something is wrong in our code, however if we return error here
-		// the NP will be retried infinitely to be added. Another option would be to fatal out here
-		// but that seems too aggressive
-		klog.Errorf("During add network policy, policy already found for %s/%s, this should not happen!",
+	// To not hold nsLock for the whole process on network policy creation, we do the following:
+	// 1. save required namespace information to use for netpol create
+	// 2. create network policy without ns Lock
+	// 3. lock namespace
+	// 4. check if namespace information related to network policy has changed, run the same function as on namespace update
+	// 5. subscribe to namespace update events
+	// 6. unlock namespace
+
+	// 1. save required namespace information to use for netpol create,
+	nsInfo, nsUnlock := oc.getNamespaceLocked(policy.Namespace, true)
+	if nsInfo == nil {
+		return fmt.Errorf("unable to get namespace %s for network policy: %s: namespace doesn't exist",
 			policy.Namespace, policy.Name)
-		return nil
 	}
-	np := NewNetworkPolicy(policy)
+	aclLogging := nsInfo.aclLogging
+	nsUnlock()
 
-	if len(nsInfo.networkPolicies) == 0 {
-		err := oc.addPolicyToDefaultPortGroups(np, &nsInfo.aclLogging)
+	// 2. create network policy without ns Lock, cleanup on failure
+	npKey := getPolicyKey(policy)
+	var np *networkPolicy
+	var err error
+
+	np, err = oc.createNetworkPolicy(policy, &aclLogging)
+	defer func() {
 		if err != nil {
-			nsUnlock()
-			return fmt.Errorf("adding network policy %s/%s failed: %v",
-				policy.Namespace, policy.Name, err)
-		}
-		defer func() {
-			if err != nil {
-				errDelete := oc.delPolicyFromDefaultPortGroups(np)
-				// rollback failed, best effort cleanup; won't add to retry mechanism since item doesn't exist in cache yet.
-				if errDelete != nil {
-					klog.Warningf("Rollback of default port groups and acls for policy: %s/%s failed, Unable to ensure namespace for network policy: error %v", policy.Namespace, policy.Name, errDelete)
-					return
+			// try to cleanup network policy straight away
+			// it will be retried later with add/delete network policy handlers if it fails
```

Comment:
how does it get retried later to delete, is that because of the logic in createNetworkPolicy that checks if the key already exists in the sync map?:
```
err := oc.networkPolicies.DoWithLock(npKey, func(npKey string) error {
		oldNP, found := oc.networkPolicies.Load(npKey)
		if found {
			cleanupErr := oc.cleanupNetworkPolicy(oldNP)
			return fmt.Errorf("cleanup for retrying nertwork policy create failed: %v", cleanupErr)
		}
```

Reply:
yeah, addNetworkPolicy will return an error, then retry logic will either retry create or handle delete if that comes.
in both cases cleanup will be called